core/src/main/java/org/apache/druid/math/expr/Parser.java

```diff
@@ -141,13 +332,135 @@ public void visit(Expr expr)
           {
             if (expr instanceof IdentifierExpr) {
               found.add(expr.toString());
+            } else if (expr instanceof LambdaExpr) {
+              LambdaExpr lambda = (LambdaExpr) expr;
+              for (String identifier : lambda.getIdentifiers()) {
+                found.remove(identifier);
+              }
             }
           }
         }
     );
     return Lists.newArrayList(found);
   }
 
+  public static Set<String> findArrayFnBindings(Expr expr)
+  {
+    final Set<String> arrayFnBindings = new LinkedHashSet<>();
+    expr.visit(new Expr.Visitor()
+    {
+      @Override
+      public void visit(Expr expr)
+      {
+        final Set<Expr> arrayArgs;
+        if (expr instanceof FunctionExpr && ((FunctionExpr) expr).function instanceof Function.ArrayFunction) {
+          FunctionExpr fnExpr = (FunctionExpr) expr;
+          Function.ArrayFunction fn = (Function.ArrayFunction) fnExpr.function;
+          arrayArgs = fn.getArrayInputs(fnExpr.args);
+        } else if (expr instanceof ApplyFunctionExpr) {
+          ApplyFunctionExpr applyExpr = (ApplyFunctionExpr) expr;
+          arrayArgs = applyExpr.function.getArrayInputs(applyExpr.argsExpr);
+        } else {
+          arrayArgs = Collections.emptySet();
+        }
+        for (Expr arg : arrayArgs) {
+          String s = getIdentifierOrCastIdentifier(arg);
+          if (s != null) {
+            arrayFnBindings.add(s);
+          }
+        }
+      }
+    });
+    return arrayFnBindings;
+  }
+
+  /**
+   * Visits all nodes of an {@link Expr}, collecting information about how {@link IdentifierExpr} are used
+   */
+  public static BindingDetails examineBindings(Expr expr)
+  {
+    final Set<String> freeVariables = new HashSet<>();
+    final Set<String> scalarVariables = new HashSet<>();
+    final Set<String> arrayVariables = new HashSet<>();
+    expr.visit(childExpr -> {
+      if (childExpr instanceof IdentifierExpr) {
+        // all identifiers are free variables ...
+        freeVariables.add(childExpr.toString());
+      } else if (childExpr instanceof LambdaExpr) {
+        // ... unless they are erased by appearing in a lambda expression's arguments because they will be bound by
+        // the apply expression that wraps the lambda
+        LambdaExpr lambda = (LambdaExpr) childExpr;
+        for (String identifier : lambda.getIdentifiers()) {
+          freeVariables.remove(identifier);
```

Comment:
Oops, yes, I introduced this bug when I added support for checking contradictory, introducing this `BindingDetails` type. Originally I handled scoping free variables to supply `findRequiredBindings` by handling [inside of `LambdaExpr` visit implementation](findRequiredBindings), so that identifiers there were intercepted and not visited so scoping issues like this didn't happen. 

Thinking about it a bit more, I think production of `BindingDetails` should be pushed down to the `Expr` implementations instead of doing it from outside with a visitor, and each should be responsible for collecting and combining correctly from their children. 

I'll refactor this to ensure scope is done correctly :+1: